### PR TITLE
[TextGeneration] Fix first token for streaming

### PR DESCRIPTION
**Pull request by dsikka**
# Summary
- We were not yielding the first token during streaming as that token is from the prompt logits/generated without having to go through the normal generation loop
- This fixes this issue but is temporary as I'd like to break up the `prep_for_generation` operator which is currently doing too much (setting up whatever we need for generation while also generating the first token). That involves a bit more work and I'd like to touch base with Damian before doing this to make sure it doesn't 
- Also, add a test to catch this case which was not covered by our test cases previously

# Testing
- Added a new test case to make sure the tokens generated during streaming match the tokens generated when streaming is turned off
- Also tested locally using the example code below

```python
from deepsparse import Pipeline

model_path = "hf:mgoin/TinyStories-1M-ds"

pipeline = Pipeline.create(
    task="text_generation",
    model_path=model_path,
    engine_type="onnxruntime",
    internal_kv_cache=False,
    continuous_batch_sizes=[2,4]
)
output = pipeline(["the dog barked", "the sun shined"], streaming=True, generation_kwargs={"max_length": 20, "num_return_sequences": 1})
for o in output:
    print(o)
```

For the following script, the output below shows the streaming fix: 

```bash
Fetching 11 files: 100%|█████████████████████████████████████████████████████████████| 11/11 [00:00<00:00, 172799.04it/s]
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 86289) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' and', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 90232) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' ran', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 93897) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' away', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 97412) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='.', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 101028) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='\n', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 105196) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='\n', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 109301) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='The', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 113550) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' dog', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 117632) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' was', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 121664) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' very', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 125852) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' happy', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 130190) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' and', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 134519) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' thanked', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 138762) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' the', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 142827) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' dog', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 146814) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='.', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 150732) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' He', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 154614) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' thanked', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 158588) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' the', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 162345) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' dog', score=None, finished=True, finished_reason='length')] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 179649) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' in', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 184380) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' the', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 188697) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' sky', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 192728) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='.', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 197083) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='\n', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 201350) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='\n', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 205580) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text='The', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 209564) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' little', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 213368) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' girl', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 217438) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' was', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 221606) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' so', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 225608) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' happy', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 229602) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' that', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 233656) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' she', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 237728) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' had', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 242139) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' been', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 246421) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' able', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 250788) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' to', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 254604) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' help', score=None, finished=False, finished_reason=None)] input_tokens=None
created=datetime.datetime(2024, 1, 17, 21, 22, 20, 258102) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' the', score=None, finished=True, finished_reason='length')] input_tokens=None
```

The above generations match what we get when we dont run streaming:
```
Fetching 11 files: 100%|█████████████████████████████████████████████████████████████| 11/11 [00:00<00:00, 146004.25it/s]
created=datetime.datetime(2024, 1, 17, 21, 22, 51, 679303) prompts=['the dog barked', 'the sun shined'] generations=[GeneratedText(text=' and ran away.\n\nThe dog was very happy and thanked the dog. He thanked the dog', score=None, finished=True, finished_reason='length'), GeneratedText(text=' in the sky.\n\nThe little girl was so happy that she had been able to help the', score=None, finished=True, finished_reason='length')] input_tokens=None
```

### CODE REVIEW

**Comment by dbogunowicz** [comment]
@dsikka let's look into it during our PP session

File: tests/deepsparse/schedulers/test_continuous_batching_scheduler.py [hunk 1]
```diff
@@ -16,10 +16,12 @@
 
 import numpy
 
+import pytest
 from deepsparse.operators import EngineOperator
 from deepsparse.schedulers import ContinuousBatchingScheduler
 
 
+@pytest.mark.skip("debuging")
```
Comment (bfineran) [inline tests/deepsparse/schedulers/test_continuous_batching_scheduler.py:24]: @dsikka revert before merge

File: tests/deepsparse/schedulers/utils/test_continuous_batching_queues.py [hunk 2]
```diff
@@ -33,6 +33,7 @@
         ([4], 5, 4),
     ],
 )
+@pytest.mark.skip("debuging")
```
Comment (bfineran) [inline tests/deepsparse/schedulers/utils/test_continuous_batching_queues.py:36]: ""

File: tests/deepsparse/transformers/pipelines/test_text_generation.py [hunk 3]
```diff
@@ -272,3 +272,13 @@ def test_streaming_with_several_prompts(pipeline, prompt):
     assert sorted(bag_of_words_first_prompt + bag_of_words_second_prompt) == sorted(
         bag_of_words_shared
     )
+
+
+def test_streaming_non_streaming_generate_same_tokens(pipeline, prompt):
```
Comment (bfineran) [inline tests/deepsparse/transformers/pipelines/test_text_generation.py:277]: great test

**Review by bfineran — APPROVED** [review]

File: src/deepsparse/operators/operator.py [hunk 4]
```diff
@@ -105,8 +105,16 @@ def __call__(
...
+        state_update = None
+        if isinstance(run_output, tuple):
+            run_output, state_update = run_output
+
         if self.has_output_schema():
-            return self.output_schema(**run_output)
+            run_output = self.output_schema(**run_output)
+
+        if state_update:
+            return run_output, state_update
```
Comment (mgoin) [inline src/deepsparse/operators/operator.py:117]: maybe the return type for this function should be updated to include this possible tuple? currently it is just `-> Any:`

**Review by bfineran — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
